SPEC.md

```diff
@@ -455,34 +670,72 @@ lines](https://www.unicode.org/versions/Unicode13.0.0/ch05.pdf):
+* `U+007F` (the Delete control character).
+* Any codepoint that is not a [Unicode Scalar
+  Value](https://unicode.org/glossary/#unicode_scalar_value).
+* `U+2066-2069`, `U+202A-202E`, `U+200E`, and `U+200F`, the [unicode
+  "direction control"
+  characters](https://www.w3.org/International/questions/qa-bidi-unicode-controls)
+
 ## Full Grammar
 
+This is the full official grammar for KDL and should be considered
+authoritative if something seems to disagree with the text above. The [grammar
+language syntax](#grammar-language) is defined below.
+
 ```
-nodes := linespace* (node nodes?)? linespace*
+document := bom? nodes
+
+nodes := (line-space* node)* line-space*
 
-node := ('/-' node-space*)? type? identifier (node-space+ node-prop-or-arg)* (node-space* node-children ws*)? node-space* node-terminator
-node-prop-or-arg := ('/-' node-space*)? (prop | value)
-node-children := ('/-' node-space*)? '{' nodes '}'
-node-space := ws* escline ws* | ws+
+plain-line-space := newline | ws | single-line-comment
+plain-node-space := ws* escline ws* | ws+
+
+line-space := plain-line-space+ ('/-' plain-node-space* node)?
+node-space := plain-node-space+ ('/-' plain-node-space* (node-prop-or-arg | node-children))?
+
+required-node-space := node-space* plain-node-space+
+optional-node-space := node-space*
+
+base-node := type? optional-node-space string (required-node-space node-prop-or-arg)* (required-node-space node-children)?
+node := base-node optional-node-space node-terminator
+final-node := base-node optional-node-space node-terminator?
+node-prop-or-arg := prop | value
+node-children := '{' nodes final-node? '}'
 node-terminator := single-line-comment | newline | ';' | eof
 
-identifier := string | bare-identifier
-bare-identifier := ((identifier-char - digit - sign) identifier-char* | sign ((identifier-char - digit) identifier-char*)?) - keyword
-identifier-char := unicode - linespace - [\/(){}<>;[]=,"]
-keyword := boolean | 'null'
-prop := identifier '=' value
-value := type? (string | number | keyword)
-type := '(' identifier ')'
-
-string := raw-string | escaped-string
-escaped-string := '"' character* '"'
-character := '\' escape | [^\"]
-escape := ["\\/bfnrt] | 'u{' hex-digit{1, 6} '}'
+prop := string optional-node-space equals-sign optional-node-space value
+value := type? optional-node-space (string | number | keyword)
+type := '(' optional-node-space string optional-node-space ')'
+equals-sign := See Table (Equals Sign)
+
+string := identifier-string | quoted-string | raw-string
+
+identifier-string := unambiguous-ident | signed-ident | dotted-ident
+unambiguous-ident := ((identifier-char - digit - sign - '.') identifier-char*) - 'true' - 'false' - 'null'
+signed-ident := sign ((identifier-char - digit - '.') identifier-char*)?
+dotted-ident := sign? '.' ((identifier-char - digit) identifier-char*)?
+identifier-char := unicode - line-space - [\\/(){};\[\]="#] - disallowed-literal-code-points
```

Comment:
`line-space` is not a character class anymore, should `- line-space` be `- newline - unicode-space`?

Reply:
The new equal signs also still need to be added (i.e. excluded).